router-tests/websocket_test.go

```diff
@@ -272,6 +280,121 @@ func TestWebSockets(t *testing.T) {
+			require.NoError(t, err)
+			initialPayload := []byte(`{"Authorization":"Bearer ` + token + `"}`)
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil, initialPayload)
+			err = conn.WriteJSON(testenv.WebSocketMessage{
+				ID:      "1",
+				Type:    "subscribe",
+				Payload: []byte(`{"query":"subscription { employeeUpdated(employeeID: 3) { id }}"}`),
+			})
+			require.NoError(t, err)
+
+			go func() {
+				xEnv.WaitForSubscriptionCount(1, time.Second*5)
+				// Trigger the subscription via NATS
+				subject := "employeeUpdated.3"
+				message := []byte(`{"id":3,"__typename": "Employee"}`)
+				err := xEnv.NatsConnectionDefault.Publish(subject, message)
+				require.NoError(t, err)
+				err = xEnv.NatsConnectionDefault.Flush()
+				require.NoError(t, err)
+			}()
+
+			var res testenv.WebSocketMessage
+			err = conn.ReadJSON(&res)
+			require.NoError(t, err)
+			require.Equal(t, "next", res.Type)
+			require.Equal(t, "1", res.ID)
+			require.JSONEq(t, `{"data":{"employeeUpdated":{"id":3}}}`, string(res.Payload))
+
+			require.NoError(t, conn.Close())
+			xEnv.WaitForSubscriptionCount(0, time.Second*5)
+		})
+	})
+	t.Run("subscription with authorization via initial payload reject", func(t *testing.T) {
+		t.Parallel()
+
+		authServer, err := jwks.NewServer(t)
+		require.NoError(t, err)
+		t.Cleanup(authServer.Close)
+		tokenDecoder, _ := authentication.NewJwksTokenDecoder(authServer.JWKSURL(), time.Second*5)
+		authOptions := authentication.WebsocketInitialPayloadAuthenticatorOptions{
+			TokenDecoder: tokenDecoder,
+			Key:          "Authorization",
+		}
+		authenticator, err := authentication.NewWebsocketInitialPayloadAuthenticator(authOptions)
+		require.NoError(t, err)
+		authenticators := []authentication.Authenticator{authenticator}
+
+		testenv.Run(t, &testenv.Config{
+			RouterOptions: []core.Option{
+				core.WithAccessController(core.NewAccessController(authenticators, true)),
+				core.WithAuthorizationConfig(&config.AuthorizationConfiguration{
+					RejectOperationIfUnauthorized: false,
+				}),
+				core.WithWebSocketConfiguration(&config.WebSocketConfiguration{
+					Enabled: true,
+					Authentication: config.WebSocketAuthenticationConfiguration{
+						FromInitialPayload: config.InitialPayloadAuthenticationConfiguration{
+							Enabled: true,
+						},
+					},
+				}),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			require.NoError(t, err)
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil, nil)
+			err = conn.WriteJSON(testenv.WebSocketMessage{
+				ID:      "1",
+				Type:    "subscribe",
+				Payload: []byte(`{"query":"subscription { employeeUpdated(employeeID: 3) { id }}"}`),
+			})
+			require.NoError(t, err)
+
+			var res testenv.WebSocketMessage
+			err = conn.ReadJSON(&res)
+			require.NoError(t, err)
+			require.Equal(t, "error", res.Type)
+			payload, err := json.Marshal(res.Payload)
+			require.NoError(t, err)
+			require.JSONEq(t, `[{"message":"unauthorized"}]`, string(payload))
```

Comment:
It won't close the ws connection. Do we want to close it?

Reply:
@StarpTech do we want to close a WS connection when auth via initial payload failed? I'd say yes, we want to close it. What's your opinion?

Reply:
The user could be able to reuse the connection for a retry in case of the IDP has changed but I think this use case is rare and we should favor better resource utilization. I also favor closing it.

Reply:
I'm a bit confused about how to update my code to correctly close the ws and also how to update my test to assert that the ws gets closed. @jensneuse or @StarpTech, can you please help me with some ideas? Thanks :)

Reply:
> The user could be able to reuse the connection for a retry in case of the IDP has changed but I think this use case is rare and we should favor better resource utilization. I also favor closing it.

Actually I think that's not correct. There can only be one single "initial payload" from the very first message sent by the client.

Reply:
> I'm a bit confused about how to update my code to correctly close the ws and also how to update my test to assert that the ws gets closed. @jensneuse or @StarpTech, can you please help me with some ideas? Thanks :)

If you don't yet have a ws conn handler, you can close it like this:
https://github.com/wundergraph/cosmo/blob/cc435bef3a20f72c03c04f54328e63dce410565c/router/core/websocket.go#L280

If you have a ws conn handler initialized already, you can close it like this:
https://github.com/wundergraph/cosmo/blob/cc435bef3a20f72c03c04f54328e63dce410565c/router/core/websocket.go#L306

In case of epoll, the conn needs to be removed:
https://github.com/wundergraph/cosmo/blob/cc435bef3a20f72c03c04f54328e63dce410565c/router/core/websocket.go#L460

In case of non-epoll, just close the conn/handler, nothing else needs to be done.

For testing, you can do multiple things.

e.g. try reading from client conn and assume you'll get an error:
https://github.com/wundergraph/cosmo/blob/cc435bef3a20f72c03c04f54328e63dce410565c/router-tests/websocket_test.go#L338

You can also wait for connection count to be zero:
https://github.com/wundergraph/cosmo/blob/cc435bef3a20f72c03c04f54328e63dce410565c/router-tests/nats_events_test.go#L84

Let me know when you think all tests pass, then I can enable the CI.
If you need further help, let me know.